### ISSUE TITLE
cond + recursion results in maximum recursion depth exceeded

### ISSUE THREAD

**Issue by BohdanK-1QBit** [google/jax#4858]
Perhaps my understanding of the mechanics of lax.cond but it is unclear to me while  the following code results in "maximum recursion depth exceeded while calling a Python object" error
```python
from jax import numpy as jnp 
from jax import lax
def diag_kron_lax(diag_operators):
    length = diag_operators.shape[0]

    def true_branch(_):
        return diag_operators.reshape(-1)
    
    def false_branch(_):
        oper1 = diag_kron_lax(diag_operators[:length//2])
        oper2 = diag_kron_lax(diag_operators[length//2:])
        return jnp.einsum('i,j->ij', oper1, oper2).reshape(-1)
    
    
    return lax.cond(length==1, true_branch, false_branch, operand=None)

diag_kron_lax(jnp.arange(10).reshape(2,-1))
```

while an equivalent pythonic version for scan (according to the documentation) works fine?
```python
from jax import numpy as jnp 
from jax import lax
def cond(pred, true_fun, false_fun, operand):
  if pred:
    return true_fun(operand)
  else:
    return false_fun(operand)

def diag_kron_my(diag_operators):
    length = diag_operators.shape[0]
    def true_branch(_):
        return diag_operators.reshape(-1)
    
    def false_branch(_):
        oper1 = diag_kron_my(diag_operators[:length//2])
        oper2 = diag_kron_my(diag_operators[length//2:])
        return jnp.einsum('i,j->ij', oper1, oper2).reshape(-1)
    
    
    return cond(length==1, true_branch, false_branch, operand=None)

diag_kron_my(jnp.arange(10).reshape(2,-1))

```

Thanks so much for any clarity on this!

**Comment by mattjj** [comment]
Thanks for the question! (You mean `cond` rather than `scan`, right?)

Unfortunately we can't stage recursive programs out of Python, for two reasons:
1. On the Python side, our tracing-based staging mechanism requires tracing (i.e. calling) both the `true_branch` and `false_branch` sides of a `lax.cond` before executing anything, so that we can see what operations are in those functions and stage them out to XLA. As a result, if either contains a recursive call the tracing will keep recursing forever. There are ways to fix that, but it won't help us because...
2. On the XLA side, XLA HLO programs can't express general recursion. Instead, they can only express a kind of [tail recursion](https://en.wikipedia.org/wiki/Tail_call) (in the [While operation](https://www.tensorflow.org/xla/operation_semantics#while), which underlies `lax.while_loop`). In principle with enough work we could set up our tracing mechanism to be able to stage out tail-recursive code, lowering it to an HLO While loop, but that wouldn't work for this case because we don't have a tail recursion (the einsum/reshape is doing some work after the recursive call).

Perhaps we could raise a better error though...

WDYT?

**Comment by BohdanK-1QBit** [comment]
thanks @mattjj for an illuminating answer! 

With your explanation about the tracing of both branches, the resulting behaviour makes sense :) I think for me, a small discussion in the documentation about this detail would have made it clear. In retrospect, it makes sense since the endless recursion was occurring with abstract tracing arrays and not their concrete values. I think, the confusing part was the difference in behaviour between lax.cond() with its pythonic version the documentation uses to illustrate lax.cond().

As for my personal implementation, would the following python's control result in a suboptimal jit'ed behaviour on accelerators given the shape of the input isn't changing (so only a single sequence in the branch tree is relevant for the computation)?

```python
from jax import numpy as jnp 
from jax import jit
def diag_kron(diag_operators):
    length = diag_operators.shape[0]
    if  length == 1:
        return diag_operators[0]
    else:
        oper1 = diag_kron(diag_operators[:length//2])
        oper2 = diag_kron(diag_operators[length//2:])
        return jnp.einsum('i,j->ij', oper1, oper2).reshape(-1) 

jit_diag_kron = jit(diag_kron)
jit_diag_kron(jnp.arange(10).reshape(2,-1))
jit_diag_kron(jnp.ones(10).reshape(2,-1))

```
P.S. I corrected the typo scan to cond :)

**Comment by mattjj** [comment]
Yes, that code looks good to me; indeed the Python control flow will effectively be unrolled in the staged-out computation, and it's fine as long as it switches on things like shape and dtype (because we trace-specialize on that information, so it's available while the Python code is executing).

Using `jax.make_jaxpr` is a quick way to check how JAX sees your function:

```python
from jax import numpy as jnp
from jax import jit, make_jaxpr
def diag_kron(diag_operators):
    length = diag_operators.shape[0]
    if  length == 1:
        return diag_operators[0]
    else:
        oper1 = diag_kron(diag_operators[:length//2])
        oper2 = diag_kron(diag_operators[length//2:])
        return jnp.einsum('i,j->ij', oper1, oper2).reshape(-1)›

jaxpr = make_jaxpr(diag_kron)(jnp.arange(10).reshape(2,-1))
print(jaxpr)
```

```
{ lambda a b c d ; e.
  let f = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] 0
      g = concatenate[ dimension=0 ] a f
      h = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0, 1), collapsed_slice_dims=(), start_index_map=(0,))
                  slice_sizes=(1, 5) ] e g
      i = broadcast_in_dim[ broadcast_dimensions=(0, 1)
                            shape=(1, 5) ] h
      j = lt 0 0
      k = add 0 1
      l = select j k 0
      m = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] l
      n = concatenate[ dimension=0 ] b m
      o = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0,), collapsed_slice_dims=(0,), start_index_map=(0,))
                  slice_sizes=(1, 5) ] i n
      p = broadcast_in_dim[ broadcast_dimensions=(0,)
                            shape=(5,) ] o
      q = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] 1
      r = concatenate[ dimension=0 ] c q
      s = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0, 1), collapsed_slice_dims=(), start_index_map=(0,))
                  slice_sizes=(1, 5) ] e r
      t = broadcast_in_dim[ broadcast_dimensions=(0, 1)
                            shape=(1, 5) ] s
      u = lt 0 0
      v = add 0 1
      w = select u v 0
      x = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] w
      y = concatenate[ dimension=0 ] d x
      z = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(0,), collapsed_slice_dims=(0,), start_index_map=(0,))
                  slice_sizes=(1, 5) ] t y
      ba = broadcast_in_dim[ broadcast_dimensions=(0,)
                             shape=(5,) ] z
      bb = xla_call[ backend=None
                     call_jaxpr={ lambda  ; a b.
                                  let c = dot_general[ dimension_numbers=(((), ()), ((), ()))
                                                       precision=None ] b a
                                      d = transpose[ permutation=(1, 0) ] c
                                  in (d,) }
                     device=None
                     donated_invars=(False, False)
                     name=_einsum ] p ba
      bc = reshape[ dimensions=None
                    new_sizes=(25,) ] bb
  in (bc,) }
```

You can also get the XLA view using `jax.xla_computation`:

```python
computation = jax.xla_computation(diag_kron)(jnp.arange(10).reshape(2, -1))
print(computation.as_hlo_text())
```

```
HloModule xla_computation_diag_kron.67

reduction_computation.41 {
  constant.44 = pred[] constant(false)
  parameter.42 = s32[] parameter(0)
  parameter.43 = s32[] parameter(1)
  ROOT add.45 = s32[] add(parameter.42, parameter.43)
}

jit__einsum.46 {
  constant.49 = pred[] constant(false)
  constant.50 = pred[] constant(false)
  parameter.48 = s32[5]{0} parameter(1)
  broadcast.51 = s32[5,1]{1,0} broadcast(parameter.48), dimensions={0}
  reshape.53 = s32[5]{0} reshape(broadcast.51)
  broadcast.54 = s32[5,5]{1,0} broadcast(reshape.53), dimensions={0}
  parameter.47 = s32[5]{0} parameter(0)
  broadcast.52 = s32[1,5]{1,0} broadcast(parameter.47), dimensions={1}
  reshape.55 = s32[5]{0} reshape(broadcast.52)
  broadcast.56 = s32[5,5]{1,0} broadcast(reshape.55), dimensions={1}
  multiply.57 = s32[5,5]{1,0} multiply(broadcast.54, broadcast.56)
  constant.58 = s32[] constant(0)
  broadcast.59 = s32[] broadcast(constant.58), dimensions={}
  reduce.60 = s32[5,5]{1,0} reduce(multiply.57, broadcast.59), dimensions={}, to_apply=reduction_computation.41
  transpose.61 = s32[5,5]{0,1} transpose(reduce.60), dimensions={1,0}
  ROOT tuple.62 = (s32[5,5]{0,1}) tuple(transpose.61)
}

ENTRY xla_computation_diag_kron.67 {
  constant.6 = pred[] constant(false)
  parameter.5 = s32[2,5]{1,0} parameter(0)
  constant.1 = s32[0]{0} constant({})
  constant.7 = s32[] constant(0)
  broadcast.8 = s32[1]{0} broadcast(constant.7), dimensions={}
  concatenate.9 = s32[1]{0} concatenate(constant.1, broadcast.8), dimensions={0}
  gather.10 = s32[1,5]{1,0} gather(parameter.5, concatenate.9), offset_dims={0,1}, collapsed_slice_dims={}, start_index_map={0}, index_vector_dim=0, slice_sizes={1,5}
  broadcast.11 = s32[1,5]{1,0} broadcast(gather.10), dimensions={0,1}
  constant.2 = s32[0]{0} constant({})
  constant.12 = s32[] constant(0)
  constant.13 = s32[] constant(0)
  compare.14 = pred[] compare(constant.12, constant.13), direction=LT, type=UNSIGNED
  constant.15 = s32[] constant(0)
  constant.16 = s32[] constant(1)
  add.17 = s32[] add(constant.15, constant.16)
  constant.18 = s32[] constant(0)
  select.19 = s32[] select(compare.14, add.17, constant.18)
  broadcast.20 = s32[1]{0} broadcast(select.19), dimensions={}
  concatenate.21 = s32[1]{0} concatenate(constant.2, broadcast.20), dimensions={0}
  gather.22 = s32[5]{0} gather(broadcast.11, concatenate.21), offset_dims={0}, collapsed_slice_dims={0}, start_index_map={0}, index_vector_dim=0, slice_sizes={1,5}
  broadcast.23 = s32[5]{0} broadcast(gather.22), dimensions={0}
  constant.3 = s32[0]{0} constant({})
  constant.24 = s32[] constant(1)
  broadcast.25 = s32[1]{0} broadcast(constant.24), dimensions={}
  concatenate.26 = s32[1]{0} concatenate(constant.3, broadcast.25), dimensions={0}
  gather.27 = s32[1,5]{1,0} gather(parameter.5, concatenate.26), offset_dims={0,1}, collapsed_slice_dims={}, start_index_map={0}, index_vector_dim=0, slice_sizes={1,5}
  broadcast.28 = s32[1,5]{1,0} broadcast(gather.27), dimensions={0,1}
  constant.4 = s32[0]{0} constant({})
  constant.29 = s32[] constant(0)
  constant.30 = s32[] constant(0)
  compare.31 = pred[] compare(constant.29, constant.30), direction=LT, type=UNSIGNED
  constant.32 = s32[] constant(0)
  constant.33 = s32[] constant(1)
  add.34 = s32[] add(constant.32, constant.33)
  constant.35 = s32[] constant(0)
  select.36 = s32[] select(compare.31, add.34, constant.35)
  broadcast.37 = s32[1]{0} broadcast(select.36), dimensions={}
  concatenate.38 = s32[1]{0} concatenate(constant.4, broadcast.37), dimensions={0}
  gather.39 = s32[5]{0} gather(broadcast.28, concatenate.38), offset_dims={0}, collapsed_slice_dims={0}, start_index_map={0}, index_vector_dim=0, slice_sizes={1,5}
  broadcast.40 = s32[5]{0} broadcast(gather.39), dimensions={0}
  call.63 = (s32[5,5]{0,1}) call(broadcast.23, broadcast.40), to_apply=jit__einsum.46
  get-tuple-element.64 = s32[5,5]{0,1} get-tuple-element(call.63), index=0
  reshape.65 = s32[25]{0} reshape(get-tuple-element.64)
  ROOT tuple.66 = (s32[25]{0}) tuple(reshape.65)
}
```

This is the un-optimized HLO. There's a way to see the backend-specific optimized HLO as well, which indicates where fusion optimizations and the like are performed, but it's not yet plumbed through to the jax Python API very nicely. I expect it will be soon!

**Comment by BohdanK-1QBit** [comment]
That is awesome for optimizing performance and debugging. Thanks again for in-depth explanations!

Although the issue has been resolved, a mystery that remains for me is that the jaxpr is almost independent on the depth of the recursion implicitly set by the length variable. I would think that it grows together with the depth of the recursion but it seems to be not the case...

**Comment by mattjj** [comment]
It may be just because in the test the leading axis is always of size 2. If I run this

```python
from jax import numpy as jnp
from jax import jit, make_jaxpr
def diag_kron(diag_operators):
    length = diag_operators.shape[0]
    if  length == 1:
        return diag_operators
    else:
        oper1 = diag_kron(diag_operators[:length//2])
        oper2 = diag_kron(diag_operators[length//2:])
        return jnp.einsum('i,j->ij', oper1, oper2).reshape(-1)

for length in [10, 100, 1000]:
  jaxpr = make_jaxpr(diag_kron)(jnp.arange(length))
  print(length, len(jaxpr.jaxpr.eqns))
```

I get this:

```
10 90
100 990
1000 9990
```

(The last one gets slow! The growth is linear rather than logarithmic because we're printing a constant number of eqns for every node in the recursion tree, not just leaves.)

I think we sorted out the main issue here, and while we could possibly generate better errors, I don't have a good idea of how to do that right now. I'll close this issue, but please don't hesitate to open more as issues/questions arise!
